testng-core/src/main/java/org/testng/TestNG.java

```diff
@@ -1083,10 +1083,14 @@ private void runSuiteAlterationListeners() {
   }
 
   private void runExecutionListeners(boolean start) {
-    for (IExecutionListener l : m_configuration.getExecutionListeners()) {
-      if (start) {
+    if (start) {
+      for (IExecutionListener l : m_configuration.getExecutionListeners()) {
         l.onExecutionStart();
-      } else {
+      }
+    } else {
+      List<IExecutionListener> executionListeners = m_configuration.getExecutionListeners();
+      Collections.reverse(executionListeners);
```

Comment:
This changes the internal state right? Should we make a copy of the listeners and then sort them ?